lib/de.ml

```diff
@@ -2605,6 +2887,1050 @@ module Def = struct
+              ; counters.(i) <- counters.(i) + 1)
+            else lens.(sym) <- 0
+          done
+        ; let counters_pos = counters.(num_counters - 2) in
+          let counters_len =
+            counters.(num_counters - 1) - counters.(num_counters - 2) in
+          let to_sort = Array.sub symout counters_pos counters_len in
+          Array.sort
+            (fun i j ->
+              match i, j with 0, _ -> 1 | _, 0 -> -1 | _ -> Int.compare i j)
+            to_sort
+          ; Array.blit to_sort 0 symout counters_pos counters_len
+          ; !num_used_syms
+
+    let build_tree a sym_count =
+      let i = ref 0 in
+      let b = ref 0 in
+      let e = ref 0 in
+      while sym_count - !e > 1 do
+        let m, n = ref 0, ref 0 in
+        if
+          !i <> sym_count
+          && (b = e || a.(!i) lsr num_symbol_bits <= a.(!b) lsr num_symbol_bits)
+        then (
+          m := !i
+          ; incr i)
+        else (
+          m := !b
+          ; incr b)
+        ; if
+            !i <> sym_count
+            && (b = e
+               || a.(!i) lsr num_symbol_bits <= a.(!b) lsr num_symbol_bits)
+          then (
+            n := !i
+            ; incr i)
+          else (
+            n := !b
+            ; incr b)
+        ; let freq_shifted =
+            (a.(!m) land lnot symbol_mask) + (a.(!n) land lnot symbol_mask)
+          in
+          a.(!m) <- a.(!m) land symbol_mask lor (!e lsl num_symbol_bits)
+          ; a.(!n) <- a.(!n) land symbol_mask lor (!e lsl num_symbol_bits)
+          ; a.(!e) <- a.(!e) land symbol_mask lor freq_shifted
+          ; incr e
+      done
+
+    let compute_length_counts a root_idx len_counts max_codeword =
+      len_counts.(1) <- 2
+      ; a.(root_idx) <- a.(root_idx) land symbol_mask
+
+      ; let rec f = function
+          | -1 -> ()
+          | node ->
+            let parent = a.(node) lsr num_symbol_bits in
+            let parent_depth = a.(parent) lsr num_symbol_bits in
+            let depth = parent_depth + 1 in
+            let len = ref depth in
+            a.(node) <- a.(node) land symbol_mask lor (depth lsl num_symbol_bits)
+            ; if !len >= max_codeword then (
+                len := max_codeword - 1
+                ; while len_counts.(!len) == 0 do
+                    decr len
+                  done)
+            ; len_counts.(!len) <- len_counts.(!len) - 1
+            ; len_counts.(!len + 1) <- len_counts.(!len + 1) + 2
+            ; f (node - 1) in
+        f (root_idx - 1)
+
+    let gen_codewords a lens len_counts max_codeword_len num_syms =
+      let next_codewords = Array.make (max_max_codeword_len + 1) 0 in
+      let rec f i len =
+        match len with
+        | 0 -> ()
+        | _ ->
+          let count = ref len_counts.(len) in
+          while !count <> 0 do
+            decr count
+            ; lens.(a.(!i) land symbol_mask) <- len
+            ; incr i
+          done
+          ; f i (len - 1) in
+      f (ref 0) max_codeword_len
```

Comment:
Hmmhmm, maybe translate the `i` argument to a simple integer instead of a `ref int` or translate the recursion to a loop with `ref`.

Reply:
Removing the recursion for a loop was the best option, so this is what I did

Reply:
But can you remove the ref?